프로필 및 커버 사진업로드 구현

dpudpu opened the pull request:
from. 철시 

현재 제 컴퓨터에서 S3에 접근이 불가해 실제로 Profile과 Cover 사진 업로드가 제대로 되는지는 확인 못했습니다. 확인 부탁드려요 ㅎㅎ

현재 Upload를 UploadType이라는 enum에서 해주고 있어요. 구조가 맞게 이뤄졌는지 여러분의 의견도 궁금합니다. *(UploadType의 이름이 적절하지 않은 것 같기도하네요...) 확인부탁드려요!

resolved #33 #68 #87

pkch93 committed:
feat: 프로필 및 커버 사진 업로드 로직 구현
- 프로필, 커버사진 업로드를 위한 서비스 클래스 생성 (UserImageUploadService)
- 프로필, 커버를 구별하는 enum UploadType
- ImageUploader 함수형 인터페이스로 업로드 구현
- 유저의 이미지를 받는 Dto UploadImage
- UploadType enum을 String에서 변환해주는 UploadTypeConverter 구현 및 등록

pkch93 committed:
feat: User에 profile과 cover 연관관계 매핑

pkch93 committed:
feat: upload url 매핑

pkch93 committed:
feat: 프로필 및 커버 사진 프론트 구현

pkch93 committed:
styles: 전체 코드 포메팅

pkch93 committed:
feat: User에 프로필, 커버 사진 매핑 로직 적용

pkch93 committed:
feat: 현재 로그인한 유저와 mypage의 유저 검증 로직 구현

pkch93 committed:
feat: 마이페이지 접근시 친구의 정보 및 내가 올린 사진 정보 추가
- 마이페이지 전용 Dto MyPageResponse 추가
- 친구 정보를 가져오는 jpql 추가 (findFriendsByUserId)
- 내가 올린 사진 정보 가져오는 쿼리 메서드 추가 (findByOwnerId)

pkch93 committed:
test: 프로덕션 코드 수정에 따른 코드 변경

pkch93 committed:
feat: mypage 프론트 작업

pkch93 committed:
Merge branch 'develop' of https://github.com/woowa-turkey/miniprojects-2019 into feature/profile

# Conflicts:
#	src/main/java/com/wootecobook/turkey/user/controller/UserApiController.java
#	src/main/java/com/wootecobook/turkey/user/service/UserService.java
#	src/main/resources/static/css/mypage.css
#	src/main/resources/static/js/mypage/mypage.js
#	src/main/resources/static/js/mypage/template.js

pkch93 committed:
fix: 로그인유저와 마이페이지 유저가 같은 경우 upload가 되지 않는 현상 해결

pkch93 committed:
test: 테스트 코드 작성
- 친구의 정보를 가져오는 JPQL 테스트
- 프로필 및 커버 사진 업로드 테스트

dpudpu committed:
fix: 프로필 사진 프론트

school0bhy commented on src/main/java/com/wootecobook/turkey/commons/converter/UploadTypeConverter.java:
```diff
@@ -0,0 +1,18 @@
+            if (type.equals(source)) {
+                return type;
+            }
+        }
```
UploadType.values().stream().filter(type -> type.equals(source)).findFirst().orElse(null)
이런식으로 스트림을 사용해서 depth를 줄일 수 있을 거 같아요!

pkch93 replied:
좋은거 같아요! 반영할께요 :)

school0bhy commented on src/main/java/com/wootecobook/turkey/file/service/UploadFileService.java:
```diff
@@ -49,4 +51,10 @@ private String getUploadPath(final MultipartFile multipartFile, final String dir
         return String.valueOf(UUID.randomUUID());
     }
+
+    public List<FileFeature> findFileFeatureByUserId(Long userId) {
```
list를 리턴하니 findFileFeatures는 어떤가요?

school0bhy commented on src/main/java/com/wootecobook/turkey/file/service/UploadFileService.java:
```diff
@@ -49,4 +51,10 @@ private String getUploadPath(final MultipartFile multipartFile, final String dir
+    public List<FileFeature> findFileFeatureByUserId(Long userId) {
+        return uploadFileRepository.findByOwnerId(userId).stream()
+                .map(UploadFile::getFileFeature)
+                .collect(Collectors.toList());
```
리턴할 때 unmodifiablelist로 리턴하는건 어떤가요?

school0bhy reviewed (approved):
너무너무 고생했어요 철시! 👍

skygl commented on src/main/java/com/wootecobook/turkey/commons/converter/UploadTypeConverter.java:
```diff
@@ -0,0 +1,18 @@
+
+    @Override
+    public UploadType convert(final String source) {
+        for (UploadType type : UploadType.values()) {
```
로직이 UploadType에 있는건 어떨까요?

pkch93 replied:
현재 UploadType을 Dto의 인자로 받는데 String에서 제가 커스텀하게 만든 UploadType을 반환하는 컨버터가 스프링에 존재하지 않아서 UploadTypeConverter를 만들고 등록했어요!

컨버터를 현재 스프링 앱에 등록하지 않으면 인자로 받을 수 없어 UploadTypeConverter를 구현했습니다 ㅎㅎ

skygl commented on src/main/java/com/wootecobook/turkey/user/service/UploadType.java:
```diff
@@ -0,0 +1,33 @@
+
+    private final ImageUploader imageUploader;
+
+    UploadType(final ImageUploader imageUploader) {
```
👍

skygl commented on src/main/java/com/wootecobook/turkey/user/service/UserImageUploadService.java:
```diff
@@ -0,0 +1,35 @@
+
+    public FileFeature uploadImage(UploadImage profile, Long userId, Long loginUserId) {
+        User user = userService.findById(userId);
+        if (!user.matchId(loginUserId)) {
```
Id 비교를 find보다 먼저 하는 것은 어떨까요?

skygl commented on src/main/java/com/wootecobook/turkey/user/service/dto/MyPageResponse.java:
```diff
@@ -0,0 +1,40 @@
+    private List<UserResponse> friends;
+
+    @Builder
+    public MyPageResponse(final UserResponse user, final FileFeature cover, final List<FileFeature> images, final List<UserResponse> friends) {
```
private으로 바꿔도 되지 않을까요?

skygl commented on src/main/java/com/wootecobook/turkey/user/service/UserService.java:
```diff
@@ -78,7 +78,15 @@ private void matchId(final Long userId, final Long sessionUserId) {
+
     public boolean existsById(final Long id) {
         return userRepository.existsById(id);
+
```
한 줄 붙이는게 좋을 것 같아요 :)

skygl reviewed (approved):
수고하셨어요 💯
프론트랑 백엔드 다 깔끔한 것 같아요~

pkch93 committed:
refactor: UploadType -> ImageType으로 변경
- UploadTypeConverter -> ImageTypeConverter 클래스 이름 변경
- ImageTypeConverter의 convert 메서드 로직 stream 적용

pkch93 committed:
refactor: List 리턴시 Collections.unmodifiableList로 반환 적용

pkch93 committed:
refactor: findFileFeatureByUserId -> findFileFeaturesByUserId

pkch93 committed:
refactor: MyPageResponse의 uri를 /api/users/{userId}/mypage로 변경
- 기존의 UserResponse를 리턴하는 /api/users/{userId}는 그대로 유지

pkch93 committed:
refactor: builder 생성자 private으로 변경

pkch93 committed:
refactor: 쿼리 실행 전에 유저간 비교 로직 수행

pkch93 committed:
styles: 전체 코드 포메팅

dpudpu commented on src/main/java/com/wootecobook/turkey/file/service/UploadFileService.java:
```diff
@@ -49,4 +52,11 @@ private String getUploadPath(final MultipartFile multipartFile, final String dir
     }
+
+    public List<FileFeature> findFileFeaturesByUserId(Long userId) {
+        return Collections.unmodifiableList(uploadFileRepository.findByOwnerId(userId)
```
👍

dpudpu commented on src/main/java/com/wootecobook/turkey/user/service/ImageType.java:
```diff
@@ -0,0 +1,33 @@
+
+public enum ImageType {
+
+    PROFILE((uploadFileService, image, user) -> {
```
👍

dpudpu commented on src/main/java/com/wootecobook/turkey/user/service/ImageType.java:
```diff
@@ -0,0 +1,33 @@
+        return profile;
+    }),
+    COVER((uploadFileService, image, user) -> {
+        UploadFile cover = uploadFileService.save(image, "cover", user);
```
"cover" 는 상수로 하면 어떨까요? 
상수로 빼면 save 메소드 안 보고도 바로 파악할 수 있을거 같아요.

dpudpu commented on src/main/java/com/wootecobook/turkey/user/service/dto/MyPageResponse.java:
```diff
@@ -0,0 +1,40 @@
+                .build();
+
+        if (user.getCover() != null) {
+            myPageResponse.cover = user.getCover().getFileFeature();
```
`user.getCover().getFileFeature();`
`user.getFileFeatureOfCover` 는 어떨까요? 
`if (user.getCover() != null)`도 없앨 수 있을거 같아요

dpudpu commented on src/main/java/com/wootecobook/turkey/user/service/dto/UserResponse.java:
```diff
@@ -14,20 +15,28 @@
                 .build();
+
+        if (user.getProfile() != null) {
+            userResponse.profile = user.getProfile().getFileFeature();
```
여기도요

pkch93 committed:
refactor: cover, profile 디렉터리명 상수화

pkch93 committed:
refactor: FileFeature 검증 로직 변경
- Optional로 null 처리
